0-module-and-module-source.md

```diff
@@ -160,83 +154,166 @@ instance.source instanceof ModuleSource;
-importer, then a new syntax can be provided to only get the `ModuleSource`:
+importer, then the program can use the `source` property of the unexecuted
+module and construct a new `Module`.
 
 ```ts
-import source from 'module.js' static source syntax;
-source instanceof ModuleSource;
-const instance = new Module(source, {
-  importHook,
-  importMeta: import.meta,
-});
+import module example from 'example.js';
+module.source instanceof ModuleSource;
+const instance = new Module(module.source, { importHook });
 const namespace = await import(instance);
 ```
 
-This is important, because it is analogous to block modules, but instead of
-inline source, it is a source that must be fetched.
+In these cases, the `ModuleSource` may benefit from the origin
+being captured in host data, such that it can be evaluated
+under a no-unsafe-eval Content-Security-Policy.
+
+### Virtual host import behavior
+
+The `Module` constructor takes an optional second argument to override the
+host-defined behavior for importing its shallow dependencies.
+The second argument is a handler object.
+The `Module` constructor eagerly captures the relevant methods
+like `importHook` and `importMetaHook`, which are all optional.
+
+```js
+class ModuleHandler {
+  importHook(specifier) {
+    // Defer to the host:
+    return import.module(specifier);
+
+    // Or, make a separate instance of the same:
+    const dependency = await import.module(specifier);
+    return new Module(dependency.source);
+
+    // Or, use a block:
+    return module {};
+
+    // Or, make a new Module from whole cloth:
+    const response = await fetch(specifier);
+    const text = await response.text();
+    const source = new ModuleSource(text);
+    return new Module(source);
+  }
+}
+
+const source = new ModuleSource(`import foo from 'foo'`);
+const module = new Module(source, new ModuleHandler());
+await import(module);
+```
+
+In this example, the `importHook` will be called once with `"foo"` and must
+return a `Module` instance by any of the available means.
+
+### Virtual Host relative import behavior
+
+The above example does not account for relative import specifiers.
+For example, to emulate a Node.js module, an absolute specifier refers
+either to a nearby package dependency or a built-in module, in order of
+precedence, but a relative specifier (starting with `"./"` or `"../"`)
+gets resolved using path math relative to the physical location of the
+containing module, the `referrer`.
+
+The `Module` instance will call its hooks as methods of the module handler
+object, so module handlers are free to carry additional metadata.
+This variation on a `ModuleHandler` carries an additional `url` property
+for import resolution and assumes an appropriate `resolve` function.
+
+```js
+class ModuleHandler {
+  async importHook(specifier) {
+    const url = resolve(specifier, this.url);
+    // Use the host's loader.
+    const module = import.module(url);
```

Comment:
Is this the real/underlying host's behavior, or would it be a virtual host if this code was itself loaded through a virtualized ModuleLoader?

Reply:
The behavior of static and dynamic `import` and `import.module` behave according to the underlying host, albeit the virtual host. All of these are captured by `importHook` of the `Module` surrounding instance if there is one.

Reply:
Might be a good place to put a `(real or virtual)` to hint at that. While it'd covered later, it may not be immediately obvious at this point of the document.